### PR TITLE
Next multi timestep monitor

### PR DESCRIPTION
**Pull request by dschwoerer**
Implementation of a multi time step monitor

Useful if you want to e.g. have a high-frequency PID controller monitor or any other monitor that changes physics (e.g. a call to EIRENE) so that your physics does not depend to strongly on the output monitor interval ...

The last commit isn't directly related, but the touching annoyed me during coding ...

### CODE REVIEW

**Comment by ZedThree** [comment]
I finally managed to go through this the other week, and I think it's a solid base for the rewrite. I got pretty much all of the features we want working using this -- in particular, using fields directly from physics models in a monitor. Not sure how e.g #343 or #321 fit in here though.

It does still need some work, but it's definitely a good place to start from :+1: 

One thing that's not very clear to me is how time sub-sampling actually works in this implementation. What exactly sets the timestep, and how/when are monitors with timesteps of < 1 called?

**Comment by dschwoerer** [comment]
[here](https://github.com/dschwoerer/BOUT-dev/blob/62b9ea13a35545b8df67851a065667fb2eb1d464/examples/monitor/monitor.cxx) is an example.

If you want I can include an example of subsampling a field ...

**Comment by ZedThree** [comment]
Could you write some documentation for how it all works? e.g how does `my_output_monitor_fast(.5)` interact with the timestep set in the input file? What sets the output frequency in the end?

**Comment by dschwoerer** [comment]
I can add an example of sub sampling, using two files.

`my_output_monitor_fast(.5)` sets the output frequency to .5 time units.
If the main output repetition time is set to 2, than this samples 4 times per time step, if the main output repetition time is set to .1, than it only samples every 5 time steps.

The output frequency of the monitor is `1/.5=2`, the output frequency of the `BOUT.dmp.*` remains unchanged by the other monitors.

**Comment by dschwoerer** [comment]
[Now an example with comments!](https://github.com/dschwoerer/BOUT-dev/blob/c1bd7abe2fc94faa2f8bd202ea5452cf51caeaf1/examples/subsampling/monitor.cxx)

I hope that helps 😃

**Comment by dschwoerer** [comment]
Concerning #343 (or #321) - it is possible to merge them, but I wasn't sure whether both would be merged, therefore I kept them separate pull requests ...

**Comment by dschwoerer** [comment]
One more thing:
I think this is useful, even if this isn't the default way of doing sub/super sampling, because:
* Allows for monitors to have states (without dirty hacks).
* Allows monitors to be called independently of the output. Useful e.g. for coupling with EIRENE

**Comment by dschwoerer** [comment]
One more thing, concerning the all in one file approach:
I don't think that this is would actually require any change.
I am not sure how this is done (haven't looked at netcdf or hdf5) but the code (if I remember correctly) doesn't handle several files, this is done by the user. Therefore the user might choose to put it in one file, if the file supports it, but this is not a choice of the monitors ...

**Comment by dschwoerer** [comment]
now with [README](https://github.com/dschwoerer/BOUT-dev/blob/e336376e0f27f5bd2b1f6aea44d9161e5a3f79bf/examples/subsampling/README.md)

File: examples/subsampling/data/BOUT.inp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+function = gauss(x-0.5, 0.3) * gauss(y-0.5,0.3)
+
+[T]
+scale=1
+function=0
+
+[Probes]
+file = "PROBES.dmp"
+floats = true
+opencloase = true # false if it should be fast
```
Comment (bendudson) [inline examples/subsampling/data/BOUT.inp]: Typo: openclose

File: examples/subsampling/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+field at a higher frequency.
+
+In this example the output frequency for the fields is set to 2.
+Therefore the fields, which are solved for, and any field that is
+added with `SAVE_REPEAT()` will be written out every second timeunit.
+
+In the code another monitor is added with an output timestep of 0.01.
+This Monitor uses a separat datafile, (via class `SimpleDatafile`, see monitor.cxx
+line 12 ff) to write the data. The options for the simple datafile are
+read fro the `[probes]` section in the input.
```
Comment (bendudson) [inline examples/subsampling/README.md]: typo: from

File: src/solver/monitor.cxx [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#include <bout/solver.hxx>
+
+// int Monitor::call(Solver * solver, BoutReal time, int iter, int nout){
```
Comment (bendudson) [inline src/solver/monitor.cxx]: Please remove this commented-out code

**Pull request merged**